pallets/fiat_filter/src/lib.rs

```diff
@@ -0,0 +1,306 @@
+    }
+    fn compute_call_fee_dock_(
+        call: &<T as Config>::Call,
+    ) -> Result<BalanceOf<T>, DispatchErrorWithPostInfo> {
+        use sp_std::convert::TryInto;
+        let fee_fiat_permill: AmountUsd = Self::get_call_fee_fiat_(call)?;
+        let fee_fiat_perbill: u32 = fee_fiat_permill.deconstruct() * 1000;
+
+        let dock_fiat_rate_permill: u32 = Self::dock_fiat_rate().deconstruct();
+        let fee_dock_permill_u64: u64 = (fee_fiat_perbill as u64 * 1_000_u64)
+            .checked_div(dock_fiat_rate_permill as u64)
+            .ok_or(Error::<T>::DivideByZero)?;
+        let fee_dock_permill_u32: u32 = fee_dock_permill_u64
+            .try_into()
+            .or(Err(Error::<T>::ArithmeticOverflow))?;
+
+        // The token has 6 decimal places (defined in the runtime)
+        // pub const DOCK: Balance = 1_000_000;
+        // T::Balance is already expressed in DOCK_Permill or microdock
+        let fee_microdock = <BalanceOf<T>>::from(fee_dock_permill_u32);
+        Ok(fee_microdock)
+    }
+
+    fn charge_fees_(who: T::AccountId, amount: BalanceOf<T>) -> Result<(), DispatchError> {
+        let _ = <T::Currency>::withdraw(
+            &who,
+            amount,
+            WithdrawReasons::FEE,
+            ExistenceRequirement::KeepAlive,
+        )?;
+        Ok(())
+    }
+
+    fn execute_call_(origin: T::Origin, call: &<T as Config>::Call) -> DispatchResultWithPostInfo {
+        let weight = call.get_dispatch_info().weight;
+        // check signature before charging any fees
+        let sender = ensure_signed(origin.clone())?;
+
+        // calculate fee based on type of call
+        let fee_dock = Self::compute_call_fee_dock_(&call)?;
+        // deduct fees based on Currency::Withdraw
+        Self::charge_fees_(sender, fee_dock)?;
+
+        let dispatch_result = call.clone().dispatch_bypass_filter(origin);
+        return match dispatch_result {
+            Ok(_post_dispatch_info) => Ok(PostDispatchInfo {
+                actual_weight: Some(weight),
+                pays_fee: Pays::No,
+            }),
+            Err(e) => {
+                frame_support::debug::info!("{:?}", &e);
+                Err(DispatchErrorWithPostInfo {
+                    post_info: PostDispatchInfo {
+                        actual_weight: Some(weight),
+                        pays_fee: Pays::No,
+                    },
+                    error: e.error,
+                })
+            }
+        };
+    }
+
+    fn set_dock_fiat_rate_(rate: Permill) {
+        // Update storage
+        DockFiatRate::put(rate);
+        // on storage updated, emit an event
+        Self::deposit_event(RawEvent::DockFiatRateUpdated(rate));
+    }
+
+    fn set_update_freq_(new_update_freq: T::BlockNumber) {
+        // Update storage
+        UpdateFreq::<T>::put(new_update_freq);
+        // Emit an event.
+        Self::deposit_event(RawEvent::UpdateFreqUpdated(new_update_freq));
+    }
+}
+
+mod amount_fiat {
```

Comment:
Do you still plan to do this? In this PR?

Reply:
Not anymore. I deleted this now.